### ISSUE TITLE
singularity install download link doesn't exist anymore

### ISSUE THREAD

**Issue by myoung3** [singularityhub/github-ci#2]
https://github.com/singularityhub/github-ci/blob/713df0b3b1d7f97aba421210ae297a343227e80e/.github/workflows/native-install.yml#L36

**Comment by vsoch** [comment]
Oh geez, yes this is  because of the repository change with singularity / sylabs - let me quickly fix this.

**Comment by vsoch** [comment]
Give this a shot? https://github.com/singularityhub/github-ci/blob/master/.github/workflows/native-install.yml

**Comment by myoung3** [comment]
Yeah I've got it working here https://github.com/kaufman-lab/github-ci/blob/master/.github/workflows/native-install.yml 

We chose different version numbers for go and singularity but it probably doesn't matter.

Any advice on this? https://github.com/kaufman-lab/github-ci/runs/3225807086?check_suite_focus=true
 "ERROR push is not implemented for docker. Why don't you add it?"

I suppose I have no affinity to dockerhub, so the easiest solution is to just use something else.

**Comment by myoung3** [comment]
fwi see that I added:

```
sudo -H pip3 install setuptools
sudo -H pip3 install wheel
```
under deploy container. `sregistry` installation was failing without them. Might want to update your readme.

**Comment by myoung3** [comment]
oh and the -H flag in `sudo -H pip3 install sregistry` seems to be necessary now

**Comment by vsoch** [comment]
Ah! So for the push part I would recommend using one of these examples to push to docker hub, quay, or github packages: https://github.com/autamus/container-builder-template. You really don't even need sregistry.

**Comment by vsoch** [comment]
@myoung3 if you have some changes that helped your deployment, please open a pull request to fix for others if you have time in the next few weeks!

**Comment by myoung3** [comment]
Thanks, will do

**Comment by myoung3** [comment]
@vsoch Any tricks on pushing a .sif to dockerhub? I thought this was theoretically possible, but maybe that was crazy. docker push doesn't seem to allow specifying a file, it can only add images recognized by docker: https://docs.docker.com/engine/reference/commandline/push/

**Comment by myoung3** [comment]
nevermind I don't know where I got that crazy idea, obviously if I'm pushing to dockerhub it needs to be a docker image

**Comment by vsoch** [comment]
Nope you unfortunately can’t do that - registries can accept custom content types like sif, but most of the production registries haven’t extended types. I think Cyverse is going with harbor and they could potentially do that.

so my suggestion is to either push a Docker container and pull down to Singularity, or do a trick to store sifs as artifacts alongside releases (proof of concept) https://github.com/singularityhub/singularity-deploy and hope that someday GitHub packages moves forward with supporting arbitrary blobs. You can technically put a SIF inside a Docker container and have a tool to extract, but it’s redundant imho.

**Comment by myoung3** [comment]
Thanks, yeah I think just building a docker image makes sense. The reason I didn't start with that technique was your caveat in the readme: 

"While the second option is faster to complete and a more simple workflow, it should be noted that docker runs with --privileged which may lead to issues with the resulting container in a non privileged situation."

Is there a reason why I can't just drop the privileged tag? The singularity image definitely needs to be able to run as a normal (non-root) user.

**Comment by vsoch** [comment]
So just to clarify - this repository entirely is just for building Singularity, and not Docker. To distinguish between the native and Docker recipes here, the difference is building a singularity container natively on the action runner vs using a Docker image with singularity inside to do the build. The privileged flag you are talking about is here https://github.com/singularityhub/github-ci/blob/f6d6c30e67c28e2bcf5b48e655a6522adcebd05a/.github/workflows/container.yml#L13 for the Docker example, and the output would still be a singularity container. If you just want a Docker container you can run Docker build right in the workflow (the autumus getting started repo has a good example).

**Comment by vsoch** [comment]
And I added that note because building with that flag can lead to wonky results that we can’t always predict, so it’s sort of a “proceed with caution” sort of deal.

**Comment by myoung3** [comment]
ahhh, ok thanks that makes sense. In the github actions workflow, prior to pushing to dockerhub I could try to spin it up the docker image in singularity to make sure it'll functioning correctly when converted to sif. 

I supposed the main downside of this workflow is that I can't specify any singularity startscript, runscript, or apps.

**Comment by myoung3** [comment]
actually this looks promising: 
https://azure.microsoft.com/en-us/blog/azure-container-registry-now-supports-singularity-image-format-containers/
https://sylabs.io/guides/3.7/user-guide/cli/singularity_pull.html

**Comment by vsoch** [comment]
That’s a good point - but actually if you just use the scientific filesystem executable you can :)

https://sci-f.github.io///tutorial-quick-start#docker

SCIF is actually a standard that is implemented in Singularity to give that functionality, but doesn’t have to depend on it. And if you want to make the SIF conversion part of your workflow, how about just adding another step after push to just pull down to singularity?

**Comment by vsoch** [comment]
Oh neat! If you get a good recipe for azure please contribute it here so other users can benefit!

**Comment by vsoch** [comment]
But yes that’s what I was talking about with respect to registries accepting custom content types. I’d expect Azure is accepting helm charts and other metadata content types too, which is cool 😎

**Comment by myoung3** [comment]
yep: https://docs.microsoft.com/en-us/azure/container-registry/container-registry-image-formats

**Comment by vsoch** [comment]
The rumors are true, love it! <3

**Comment by myoung3** [comment]
Unfortunately it seems like the azure containers have a few limitations;

- there's no free tier (although they do offer a 12 month free trial at the standard level).
- authentication seems to be required to pull (no public registries)
- based on the interface on zure, it seems like generating tokens for the container repo might require a premium feature for the azure container repository? Not 100% sure on that. 
- singularity doc for authenticating with oras is entirely lacking. 
  - It seems admins can generate a token for themself, but it's not clear how to use these with singularity pull/push 
  - As far as I can tell, you might need to install oras to do authentication? It doesn't seem to be built into singularity.

**Comment by vsoch** [comment]
Ack, that's frustrating! For the last point, however, I do think singularity push has an `oras://` endpoint: https://sylabs.io/guides/3.5/user-guide/cli/singularity_push.html#synopsis. I've never used it, but I think that could be something to try?

**Comment by vsoch** [comment]
But actually I think you already may be trying that. It would be worth joining the SingularityCE slack and asking about how to do this.

**Comment by myoung3** [comment]
yeah singularity push definitely has an oras endpoint listed in the documentation, but it's not clear how to authenticate that push. I'm guessing authentication might need to be set up prior to `singularity push` using an oras command? Or else only unauthenticated oras pushes are currently supported.  I've decided to just use dockerhub at least for now since having an easy to pull (public) repo is more important than having singularity runscripts/apps. Hopefully the landscape changes someday and there's more support for .sif images in public repos.  Thanks so much for your help with this!

**Comment by myoung3** [comment]
fyi: github container registry apparently supports oci? 
https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry.

also I submitted a documentation issue to sylabs: https://github.com/sylabs/singularity/issues/208

**Comment by vsoch** [comment]
Yes you can push docker images to the registry, check out one of my co-projects Autamus! https://autamus.io

**Comment by myoung3** [comment]
It seems that uploading and downloading .sifs to oras://ghcr.io works.

try this:
```
singularity remote login --username <username> oras://ghcr.io
```
You'll be prompted for a PAT. Then you can use singularity push and singularity pull to `oras://ghcr.io/<registry>:<tag>`

**Comment by vsoch** [comment]
No way! We should try a build recipe that deploys there (and then gives instructions to pull)

**Comment by myoung3** [comment]
Want me to submit a PR with a new github workflow?

**Comment by myoung3** [comment]
I don't really understand how matrices work, so I'll submit something simple and you can modify it as you like

**Comment by myoung3** [comment]
actually it's probably easiest if I just modify an existing workflow and deploy twice, once to dockerhub and once to github packages.

**Comment by vsoch** [comment]
Whatever works for you! I'm having dinner now but am going to play around with this after. Super exciting if it can be an option!

**Comment by vsoch** [comment]
hey I'm working on this so don't rush to try and do it! I will need to test deploying to the package registry from CI, so I'm going to tweak the recipes and update docs for doing that.

**Comment by myoung3** [comment]
I'm testing it here: 
https://github.com/kaufman-lab/build_containers

Weirdly I can't get `singularity remote login` commands that work locally to work in the build environment.

**Comment by vsoch** [comment]
That's funny! I just hit that error too. https://github.com/singularityhub/github-ci/pull/3/checks?check_run_id=3237348014 But I haven't tried debugging yet.

**Comment by vsoch** [comment]
omg success! https://github.com/singularityhub/github-ci/pkgs/container/github-ci

I'll figure out how to pull, update the docs, and we can get this merged soon. I did a couple of examples for automated builds, all with matrices:

 - discovering changed files, build with docker base
 - discovering changed files, build natively
 - manually specified matrix, build natively

so hopefully there should be something for everyone!

**Comment by myoung3** [comment]
hmm i was trying to pipe into password-stdin as well, but I was quoting the secret which apparently was something weird to happen (possibly getting substituted to empty?). quoting locally didn't cause any problems.  anyway my login is working now too.  I also might try using this random person's actions to add singularity binaries https://github.com/marketplace/actions/setup-singularity to speed up build time.

**Comment by myoung3** [comment]
hah and now I see you made an issue in that repo like an hour ago.

**Comment by vsoch** [comment]
I tried that action with 3.8.0, but there was an error with mconfig (check out the issues on that repo) so I'm going to leave as is for now.

**Comment by vsoch** [comment]
lol and you already noticed my issue! We are like, running around the same bush like cartoon characters and telling one another what's on the other side!

**Comment by myoung3** [comment]
haha yeah the synchronicity is getting absurd! super glad to have someone working through this with me though!

**Comment by myoung3** [comment]
Are you thinking that it would be nice if we could edit or at least the `docker pull` instructions? https://github.com/kaufman-lab/build_containers/pkgs/container/geospatial_plus

Not even sure where we might submit an issue for that...

**Comment by vsoch** [comment]
Woohoo congrats on your container!

So for editing packages, the most you can do to make them private/public is click on Package settings, but as far as I know you can't edit that particular line. I think at best you can update the README.md, which is directly below that box.

**Comment by vsoch** [comment]
It is quite funny though!

```bash
$ docker pull ghcr.io/singularityhub/github-ci:latest
latest: Pulling from singularityhub/github-ci
bb3948bc2682: Pulling fs layer 
unexpected end of JSON input
```
Yeah we should tweet someone at GitHub about this.

**Comment by vsoch** [comment]
okay gave it a shot! https://twitter.com/vsoch/status/1422777308146847746.

**Comment by myoung3** [comment]
fyi it seems like anyone who wants to `singularity pull` an image needs to first load a github PAT into singularity, even for public repositories.

**Comment by vsoch** [comment]
Do you want to PR to update docs?

**Comment by myoung3** [comment]
nevermind! You can totally do anonymous pulls using singularity. I think the issue is that if you try to pull with an disabled/expired github PAT, singularity will error out rather than try again without a key.

**Comment by myoung3** [comment]
issue filed here: https://github.com/sylabs/singularity/issues/212
